aerogear/aerogear-js-sdk
# POC: JS SDK in single package for React Native (and Cordova)

## Description
It is not possible to have a single, multiplatform JS SDK since it is necessary to access to some native functionality and React Native and Cordova need different and specific implementation for that – React Native needs a [native module](https://facebook.github.io/react-native/docs/native-modules-ios.html) whereas Cordova uses [plugins](https://cordova.apache.org/docs/en/latest/guide/hybrid/plugins/).

The idea so far was to have a common library `aerogear-js-sdk` that would be imported by the user alongside the platform specific one, if needed.

This POC presents an upside-down approach: user imports a single platform-specific library, `aerogear-react-native-sdk` or `areogear-cordova-sdk`, the same exact way that it would work for the Android or iOS SDK. These platform-specific libraries contain all the native bits and they depend on a third and common library `aerogear-js-sdk` that contains JS-only common business logic.

In this POC, the Metrics service has been implemented. The React Native module contains the logic that grabs device and app information and internally uses the JS SDK core module, that sends this data  to the backend app-metrics instance.

## Points of discussion
- What's `platform`? Is it `"android"` or `"android-react"` for instance?
- What's `sdkVersion`? Is it `js-sdk` version or the platform-specific libraries `react-native-sdk`'s?
- Is the base library going to be private? In other words, will `js-sdk` package be able to be downloaded and used standalone by any JS app? Like a web app.
- How are we going to organize the repos? Are we going to have different repos for aerogear-js, react-native-sdk, aerogear-cordova-sdk...?

## TODO
- [x] JS Base SDK (metrics)
- [x] React Native SDK POC (metrics)
- [x] React Native example app implementation
- [ ] React Native SDK in TS
- [ ] React Native SDK test suite
- [x] Cordova SDK POC (metrics)
- [ ] Cordova example app implementation
- [ ] Cordova SDK in TS
- [ ] Cordova SDK test suite

## Usage / Verification
Note that any package has been published yet so dependencies has to be resolved locally and some amount of manual work is required.

1. Since JS SDK's is written in TS, it has to be transpiled in order to RN module to consume it.

```sh
cd aerogear-js-sdk/packages/core
npm install && npm run build
```

2. Clone the RN SDK: https://github.com/josemigallas/aerogear-react-native-sdk at the same level as `aerogear-js-sdk` repo is cloned. Then install dependencies:
```sh
npm install
```

4. Deploy an instance of `metrics-apb` in Openshift and copy its app-metrics URL (more info at https://github.com/aerogear/mobile-core)

5. Add a `mobile-services.json` file in the root folder (same level as package.json) with the correct metrics service configuration. Only `url` and `type` are actually relevant at this point. Example:
```json
{
  "version": 1,
  "clusterName": "",
  "namespace": "",
  "clientId": "",
  "services": [
    {
      "id": "metrics",
      "name": "metrics",
      "type": "metrics",
      "url": "http://aerogear-app-metrics-test.192.168.37.1.nip.io/metrics",
      "config": {}
    }
  ]
}
```

Note that before running on Android, an emulator must be running or a device connected. When running iOS, simulator app should start up automatically.
